Modules/mmapmodule.c

```diff
@@ -682,6 +682,80 @@ mmap__exit__method(PyObject *self, PyObject *args)
     return _PyObject_CallMethodId(self, &PyId_close, NULL);
 }
 
+static PyObject *
+mmap__repr__method(PyObject *self)
+{
+    PyObject *repr;
+    mmap_object *m_obj;
+    m_obj = (mmap_object *) self;
+
+    const char *reprfmt;
+    const char *access_str;
+
+    switch (m_obj->access) {
+        case ACCESS_DEFAULT:
+            access_str = "ACCESS_DEFAULT";
+            break;
+
+        case ACCESS_READ:
+            access_str = "ACCESS_READ";
+            break;
+
+        case ACCESS_WRITE:
+            access_str = "ACCESS_WRITE";
+            break;
+
+        case ACCESS_COPY:
+            access_str = "ACCESS_COPY";
+            break;
+
+        default:
+            // should not get here
+            assert(0);
+            return NULL;
+    }
+
+    Py_ssize_t size = m_obj->size;
+    const char *tp_name = Py_TYPE(self)->tp_name;
+
+
+#ifdef MS_WINDOWS
+    int fileno = m_obj->file_handle;
+    if (m_obj->map_handle == NULL)
+#elif defined(UNIX)
+    int fileno = m_obj->fd;
+    if (m_obj->data == NULL)
+#endif
+    {
+        reprfmt = "<%s closed=True fileno=%d access=%s>";
+        repr = PyUnicode_FromFormat(reprfmt, tp_name, fileno, access_str);
+    }
+    else {
+        const char *data = m_obj->data;
+        Py_ssize_t pos = m_obj->pos;
+
+        PyObject *length = PyLong_FromSize_t(size);
+        PyObject *offset = PyLong_FromSize_t(pos);
+
+        if (size < 64) {
+            reprfmt = "<%s closed=False fileno=%d access=%s size=%R "
+                      "offset=%R>";
+
+            repr = PyUnicode_FromFormat(reprfmt, tp_name, fileno, access_str,
+                                        length, offset);
+        }
```

Comment:
Hmm, no difference between these. I'd suggest make `closed` a local variable and the logic could be clearer.

Reply:
There is a difference, that when `data`  is not available, getting `offset` could be insecure.

Reply:
Sorry, I don't get it. Why could it be insecure? After `new_mmap_object` they are all initialized right? What I am missing here? And if they are secure, I don't think we need to distingiush them between closed and open status, just show whatever they are. One more thing, why leave out the actual `offset` property and make `pos` the `offset`? It's confusing.

Reply:
First, I'd say sorry for `offset` should be `pos`.

At the first beginning I mean if `mmap` gets closed,  the `data` of `mmap_object` has already been `NULL`. In this situation why should we allow uers to get a few outdated members like `pos` or `size`? I know people might feel like to ask a closed `mmap` for the historical records so I think you're right in that case, but in fact, the `tell` method and `size` method just raise exception `PyErr_SetString(PyExc_ValueError, "mmap closed or invalid")`:
- tell: https://github.com/python/cpython/blob/master/Modules/mmapmodule.c#L567
- size: https://github.com/python/cpython/blob/master/Modules/mmapmodule.c#L417

So there's no way to get `size` and `pos` from a `mmap` object when it has been closed, why should we break this in just `repr` method? That's also what now motivates me to still distinguish closed status from the open one.

Reply:
Hmm, I know why I once used `offset` now..
It's the argument of mmap constructor: https://github.com/python/cpython/blob/master/Modules/mmapmodule.c#L1058

Also, `length` is used there instead of `size`.

I think we should do more to make it consistent in both low level(C data structures) and high level(interfaces for  Python). 
It might be better to change `size` to `length` and consist using `offset` via 
- https://github.com/python/cpython/blob/master/Modules/mmapmodule.c#L1144
- https://github.com/python/cpython/blob/master/Modules/mmapmodule.c#L1133

The wording seems to be a thing, e.t.c, in previous codes there're already phrases like `length of mmap`, `file size` or `memory offset`:
https://github.com/python/cpython/blob/master/Modules/mmapmodule.c#L1133-L1145

Reply:
Makes sense. Some properties are meaningless when closed, we don't need them, I mean `pos`. For others, I'd suggest just using the constructor arguments, `length` is `length`, `pos` is `pos`, `offset` is `offset`. Don't use `size` please since it's confusing, it's stored internally as a caculated field but unforunately there is also a `size()` method returns something different. As for `fileno`, I am okay to omit it for now.